src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/ClassReader.java

```diff
@@ -590,7 +590,21 @@ Type classSigToType() {
                 ClassSymbol t = enterClass(readName(signatureBuffer,
                                                          startSbp,
                                                          sbp - startSbp));
-                outer = new ClassType(outer, sigToTypes('>'), t) {
+                List<Type> actuals = sigToTypes('>');
+                List<Type> formals = ((ClassType)t.type.tsym.type).typarams_field;
+                if (formals != null) {
+                    if (actuals.isEmpty())
+                        actuals = formals;
+                    else
+                        updateBounds(actuals, formals);
```

Comment:
Is it important we do this here - or can we always do this in the completer (to simplify the code some more)

Reply:
thanks for the suggestion, will give it a try